prow/tide/tide.go

```diff
@@ -1410,15 +1403,14 @@ func (c *Controller) dividePool(pool map[string]PullRequest, pjs []prowapi.ProwJ
 		}
 		sps[fn].prs = append(sps[fn].prs, pr)
 	}
-	for _, pj := range pjs {
-		if pj.Spec.Type != prowapi.PresubmitJob && pj.Spec.Type != prowapi.BatchJob {
-			continue
-		}
-		fn := poolKey(pj.Spec.Refs.Org, pj.Spec.Refs.Repo, pj.Spec.Refs.BaseRef)
-		if sps[fn] == nil || pj.Spec.Refs.BaseSHA != sps[fn].sha {
-			continue
+
+	for subpoolkey, sp := range sps {
+		pjs := &prowapi.ProwJobList{}
+		if err := c.prowJobClient.List(c.ctx, pjs, ctrlruntimeclient.MatchingField(BaseSHAIndexKey, sp.sha)); err != nil {
+			return nil, fmt.Errorf("failed to list jobs for subpool %s: %v", subpoolkey, err)
 		}
-		sps[fn].pjs = append(sps[fn].pjs, pj)
+		c.logger.WithField("subpool", subpoolkey).Infof("Found %d prowjobs.", len(pjs.Items))
+		sps[subpoolkey].pjs = pjs.Items
```

Comment:
The fact that the base SHA matches is not enough to know that this PJ matches the tide pool. It is possible for the same base sha to exist in different org/repo. More realistically, it is also possible for two different branches (`BaseRef`) to point to the same commit and we need to distinguish the different pools in this case.

Reply:
We don't necessarily need to make the index more complicated, but if we keep it as is we need to filter out PJs for other subpools like we did here before.

Reply:
Do we have unit tests for these scenarios? That's the best way to ensure it remains invariant.

Reply:
Alvaro added unit tests, but they require the `envtest` package I asked about here: https://github.com/kubernetes/test-infra/pull/14830#pullrequestreview-304852708

(Also if this is what we want to do, do you have any recommendations on the best way to do this with bazel?)

Reply:
Can we add actual unit tests? I don't see any reason why we would need etcd in order to validate the behavior of
* the same SHA existing in multiple repos
* two branches pointing to the same commit

Reply:
> Can we add actual unit tests? I don't see any reason why we would need etcd in order to validate the behavior of [...]

We have tests for "does the index func do what we expect it to do":  https://github.com/kubernetes/test-infra/pull/14830/files#diff-a71b130258acf1404ae655d199098e80R3102

What we do not have without the envtest-based integration test is testing for:
* Do we actually add the index to the cache (The only way to figure this out is by using it, that will yield an error if the given index does not exist)
* If we request an index iN, do we get the expected result (or something different from iN+1 for that matter)

There is no fake implementation of a `mgr` at all and no fake implementation of a ctrlruntime.Client that supports indexing. The only way I see of testing this in unittests is by writing a fake `mgr` and a fake `ctrlruntimeclient.Client` that supports indexing (could probably done by wrapping the existing implementation and just call the index func on list calls that contain a `matchinField` opt)

Reply:
We validate that "input X return exactly Y". This is different than "ensure input X and X' return whatever values so long as they differ from each other", right?

Aka there's nothing in that unit test hinting that when I come around in 2021 and improve this method I need to ensure that the same commit on two branches returns a unique key for each branch.

Reply:
> We validate that "input X return exactly Y". This is different than "ensure input X and X' return whatever values so long as they differ from each other", right?

Added a `TestCacheIndexFuncReturnsDifferentResultsForDifferentInputs` that verifies that.

Generally, using a cache with an index is probably the most efficient solution for filtering kube api objects repeatedly, so it would be great if we can find a way to test that that is agreeable for everyone :) How do you feel about using envtest for that @fejta ?

Reply:
Faking the interface seems like the most appropriate strategy here.

Making everything depend on making real calls to the apiserver is an inefficient test strategy. We're just consumers of all this code. Why can't we just assume kubernetes works correctly?

We do not, for example, validate the the Google cloud storage client transfers files correctly. We just assume it works and fake the return values.

I would recommend this faking strategy here over replacing unit testing with integration testing.

Reply:
Okay. I've also opened https://github.com/kubernetes-sigs/controller-runtime/issues/657 for this, because ideally we don't have to build fakes downstream.

While a test for "do we actually add the index to the cache" definitely makes sense, adding a fake client that uses an index pretty much only tests our dependencies (that do have tests for this) and our fake implementation. For the sake of getting this done, I'll add it nonetheless.

Reply:
I agree that is probably more work than is justified? This is the call to test:

```
		err := c.prowJobClient.List(
			c.ctx,
			pjs,
			ctrlruntimeclient.MatchingField(cacheIndexName, cacheIndexKey(sp.org, sp.repo, sp.branch, sp.sha)),
			ctrlruntimeclient.InNamespace(c.config().ProwJobNamespace))
```

So wrap it in a function:
```golang

type lister interface {
  List(context.Context, *prowapi.ProwJobList, matchingFieldRetValue, namespaceRetValue) error
}

func listMatchingJobs(ctx context.Context, prowJobClient lister, sp subpool, namespace string) (*prowapi.ProwJobList, error) {
  var pjs prowapi.ProwJobList
  if err := prowJobClient.List(ctx, &pjs, MatchingField(cacheIndexName, cacheIndexKey(sp.org, sp.repo, sp.branch, sp.sha)), InNamespace(namespace); err != nil {
    return nil, err
  }
  return &pjs
```
 
Now write a `fakeLister` and validate that a) we send cacheIndexName, cacheIndexKey and namespace correctly. 

Change all the list calls to this helper function.

Done.

Reply:
Well I ended up writing a small implementation of the subset of the `manager` we need and something that just embedds a an upstream fakeclient and wraps its `List` call to use an index func if requested.

I think this should be good now, PTAL.